core/commands/dag/dag.go

```diff
@@ -241,3 +277,487 @@ var DagResolveCmd = &cmds.Command{
+	// similar to pinner.Pin/pinner.Flush
+	batch := ipld.NewBatch(req.Context, (*api).Dag())
+
+	roots := make(expectedRootsSeen)
+
+	it := req.Files.Entries()
+	for it.Next() {
+
+		file := files.FileFromEntry(it)
+		if file == nil {
+			ret <- importResult{err: errors.New("expected a file handle")}
+			return
+		}
+
+		car, err := gocar.NewCarReader(file)
+		if err != nil {
+			ret <- importResult{err: err}
+			return
+		}
+
+		// Be explicit here, until the spec is finished
+		if car.Header.Version != 1 {
+			ret <- importResult{err: errors.New("only car files version 1 supported at present")}
+			return
+		}
+
+		for _, c := range car.Header.Roots {
+			if _, exists := roots[c]; !exists {
+				roots[c] = false
+				counts.RootsDeclared++
+			}
+		}
+
+		for {
+			block, err := car.Next()
+			if err != nil && err != io.EOF {
+				ret <- importResult{err: err}
+				return
+			} else if block == nil {
+				break
+			}
+
+			// the double-decode is suboptimal, but we need it for batching
+			nd, err := ipld.Decode(block)
+			if err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			if err := batch.Add(req.Context, nd); err != nil {
+				ret <- importResult{err: err}
+				return
+			}
+
+			counts.Blocks++
+
+			// encountered something known to be a root, for the first time
+			if seen, exists := roots[nd.Cid()]; exists && !seen {
+				roots[nd.Cid()] = true
+				counts.RootsSeen++
+			}
+		}
+
+		// every single file in it. is already open before we start
+		// just close here sooner rather than later for neatness
+		// this won't/can't help with not running out of handles
+		file.Close()
+	}
+
+	if err := it.Err(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	if err := batch.Commit(); err != nil {
+		ret <- importResult{err: err}
+		return
+	}
+
+	ret <- importResult{roots: &roots}
+}
+
+var DagExportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Streams the selected DAG as a .car stream on stdout.",
+		ShortDescription: `
+'ipfs dag export' fetches a dag and streams it out as a well-formed .car file.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("selector", true, false, "Selector representing the dag to export").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI. Defaults to true when STDERR is a TTY."),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		// enable progress implicitly if a TTY
+		if _, specified := req.Options[progressOptionName]; !specified {
+			errStat, _ := os.Stderr.Stat()
+			if 0 != (errStat.Mode() & os.ModeCharDevice) {
+				req.Options[progressOptionName] = true
+			}
+		}
+
+		return nil
+	},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		c, err := cid.Decode(req.Arguments[0])
+		if err != nil {
+			return fmt.Errorf(
+				"unable to parse selector (currently only bare CIDs are supported): %s",
```

Comment:
What's the format of this argument when proper selectors are allowed? I assume it'll allow parsing of @creationix's new string-form selectors at some point, but aren't we still going to need to specify a root _separate_ to the selector? Is that going to need some extra syntax to squish CID+selector together or will this need to be 2 arguments, one for CID and one for selector? And in that case, should this argument just be renamed from "selector" to "root" for now with the expected addition of a selector argument to come later?

Reply:
See also https://github.com/ipld/go-car/blob/3e2cf7af0fab7d59ac0eb2ebc3dcad1c06fa68d0/selectivecar.go#L21-L24 where these two things are distinct.

Reply:
:+1: I advocate for a tuple here.

Selectors are a declarative document which, when interpreted, navigates from one starting node, to a series of reached nodes (being all those visited) and matched nodes (being those which the selector... selects).

Starting a selector at a particular CID is a common operation -- great; it composes cleanly with the above.

Reply:
Addressed as follows: https://github.com/ipfs/go-ipfs/pull/7036/files#diff-15172926c4147422f472139313467179R254-R263

The error text: https://github.com/ipfs/go-ipfs/pull/7036/files#diff-15172926c4147422f472139313467179R269

This should alleviate the duality ( @warpfork , @rvagg  please confirm )